Extend connectivity-check for HTTP policy validation via CUE

joestringer opened the pull request:
This PR introduces a [CUE](https://cuelang.org/)-based framework for defining connectivity checks to consolidate the definitions of the various connectivity checks and support simpler, more concise declarations of new connectivity checks. 

Review commit-by-commit.

### Why CUE?

CUE is a data constraint language built defined as a superset of JSON which aims to "simplify tasks involving defining and using data".  See the second commit for a more detailed discussion, talking points below as a summary:

    In context of the connectivity check YAMLs, CUE is useful to allow us to
    "evaporate" the boilerplate necessary to define Kubernetes YAMLs for
    Deployments, Services and CiliumNetworkPolicies and allow developers to
    specify various permutations for connectivity checks concisely.

    Why should we use it?
    * It's more concise. [...]
    * Consolidate constant declaration. [...]
    * More structured data declarations for k8s resources of checks [...]. Can build powerful tooilng around this.

    What are the dangers (+mitigations)?
    * It's relatively imature, v0.2.2. This PR pins CUE usage to this version. [...]
    * Another language in the tree. [...] This however aligns with Go and k8s community exploration of the language.
    * It's easy to hide all complexity, templating away all parameters in a way that makes the declarations implicit. It'll be up to us to figure out how explicitly to declare our check definitions. [...]
    * Additive declarations. Evaluation of language statements are not ordered by line ordering. Can be counter-intuitive. [...]
 
Based on the above and the experience of converting these checks over to CUE, I believe this is a useful and powerful tool for declaring YAML variations, perfect for a use case like connectivity-checks. We may in future even be able to explore using CUE for defining manifests in the CI.

### How do we know these changes are correct?

This PR was validated against the existing "hostport" connectivity check YAML from the tree using the `make inspect` target introduced in this PR. I've attached the diff for existing checks below for reference; the commit message for the relevant changes also highlights the specifics of what has generally changed:

     * All containers are converted to use readinessProbe and not
       livenessProbe.
     * All readiness probes now specify --connect-timeout of 5s.
     * Readiness probes access `/public` or `/private` per the underlying
       container HTTP server paths rather than just accessing `/`.
     * DNS allow policies are converted to consistently allow both TCP and
       UDP-based DNS.
     * Container names are derived from pod names.
     * The new YAMLs declare additional labels for all resourcess, such as
       'component' and 'topology'.

[connectivity-check-hostport.yaml.diff.txt](https://github.com/cilium/cilium/files/4973710/connectivity-check-hostport.yaml.diff.txt)

Tested on a 2-node GKE cluster per the Cilium getting-started guide.

### Extending the connectivity checks for HTTP policy validation

See the commit log for details:

    These new checks configure various L7 proxy paths to validate
    connectivity via L7 proxies, in the following dimensions:

    - Apply policy on egress; ingress; or both (proxy-to-proxy)
    - Intra-node / Multi-node
    - Allow / Deny

    Note that proxy-to-proxy always configures egress allow policy to ensure
    that the traffic goes via the proxy and in the drop case the requests
    are only rejected at the destination. This is because applying egress
    deny at the source would prevent proxy-to-proxy connectivity, meaning
    the test would be equivalent to the egress-only reject policy case. This
    way, we ensure that the path via the egress proxy to the destination is
    tested in the reject case.

    These are implemented partially through a new 'echo-c' pod which always
    has ingress policy applied to allow GET requests to '/public'. Depending
    on whether ingress policy is needed to check the particular permutation
    the new checks may connect to 'echo-a' or 'echo-c'.

    These are implemented by adding pods for each permutation of policy
    apply point and topology; then by adding allow / deny containers within
    that pod to test the allow/deny cases.

    The 'connectivity-check-proxy.yaml' includes all of the above.

    Finally, the omissions: This commit does not attempt to address
    variations in datapath configuration. This includes IPv4 vs. IPv6;
    tunnel/direct-routing; endpoint config; kube proxy/free; encryption.
    These are left up to the cluster operator configuring Cilium in specific
    modes and subsequently deploying these YAMLs.

### Tooling

One interesting aspect around CUE is the ability to write tailored CLIs against the data to filter the data you want and output in the format you're after. See the commit message for more details of the new tooling:

      $ cue cmd help
      List connectivity-check resources specified in this directory

      Usage:
        cue [-t component=<component>] [-t name=<name>] [-t topology=<topology>] <command>

      Available Commands:
        dump   Generate connectivity-check YAMLs from the cuelang scripts
        ls     List connectivity-check resources specified in this directory

    List available connectivity-check components:

      $ cue cmd ls
      KIND                  COMPONENT        TOPOLOGY     NAME
      Service               network-check    any          echo-a
      Service               services-check   any          echo-b
      Service               services-check   any          echo-b-headless
      Service               services-check   any          echo-b-host-headless
      Deployment            network-check    any          echo-a
      Deployment            services-check   any          echo-b
      Deployment            services-check   any          echo-b-host
      Deployment            network-check    any          pod-to-a
      Deployment            network-check    any          pod-to-external-1111
      Deployment            policy-check     any          pod-to-a-allowed-cnp
      Deployment            policy-check     any          pod-to-a-denied-cnp
      Deployment            policy-check     any          pod-to-external-fqdn-allow-google-cnp
      Deployment            services-check   multi-node   pod-to-b-multi-node-clusterip
      Deployment            services-check   multi-node   pod-to-b-multi-node-headless
      Deployment            services-check   intra-node   pod-to-b-intra-node-clusterip
      Deployment            services-check   intra-node   pod-to-b-intra-node-headless
      Deployment            services-check   multi-node   host-to-b-multi-node-clusterip
      Deployment            services-check   multi-node   host-to-b-multi-node-headless
      CiliumNetworkPolicy   policy-check     any          pod-to-a-allowed-cnp
      CiliumNetworkPolicy   policy-check     any          pod-to-a-denied-cnp
      CiliumNetworkPolicy   policy-check     any          pod-to-external-fqdn-allow-google-cnp

    These can be filtered by component, topology or name. For example:

      $ cue cmd -t component=network ls
      KIND         COMPONENT       TOPOLOGY    NAME
      Service      network-check   any         echo-a
      Deployment   network-check   any         echo-a
      Deployment   network-check   any         pod-to-a
      Deployment   network-check   any         pod-to-external-1111

    Finally, to gather the (filtered) YAMLs for the specified resources:

      $ cue cmd dump | head -n 20
      metadata:
        name: echo-a
        labels:
          name: echo-a
          topology: any
          component: network-check
      spec:
        ports:
          - port: 80
        selector:
          name: echo-a
        type: ClusterIP
      apiVersion: v1
      kind: Service
      ---
      ...

#### Q: Do I need to install CUE?

No, the Makefile introduces usage of the cue docker image to run the YAML generation commands. For more advanced usage like the tooling described above however, you would need to either install CUE or run it via the docker container in a similar way to how the makefile does.

### Outstanding questions

* For now I have introduced the new checks (proxy-based and hostport+proxy which are quarantined) into separate YAMLs to allow them to be tested independently. They are not part of the simple YAML. This allows the regular connectivity-check YAML to remain small and contained to allow it to reliably pass in smaller environments. We should consider how to expand the testing of these YAMLs in CI.

Related: #8491

joestringer committed:
connectivity-check: Add 'make clean' support

Factor out the targets for all YAMLs so it can be reused by a new phony
target, 'clean'.

Signed-off-by: Joe Stringer <joe@cilium.io>

jrajahalme reviewed (approved):
The general approach looks good, did not really review the yamls in detail. There are some additional dimensions that have been problematic in the past:

- Service access via hostport/not
- Hostport backend POD in the same v.s. other node as the one that was addressed.

Then there are bunch of datapath config dimensions (tunnel/no tunnel, endpoint routes, kube proxy/free, encryption/no, etc. ), but these are out of scope for the connectivity checks that apply to any given config.

kkourt commented on examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml:
```diff
@@ -0,0 +1,32 @@
+        imagePullPolicy: IfNotPresent
+        livenessProbe:
+          exec:
+            command: ["ash", "-c", "! curl -sS --fail --connect-timeout 5 -o /dev/null echo-c/invalid-path"]
```
I wonder if it would be better to check both allow and reject in a single yaml file since it is the same policy. I.e., do something like: `curl ... echo-c/ && ! curl ... echo-c/invalid-path`

Nit: `invalid-path` was a bit confusing for me because it was not clear that it would be served by the server. Maybe `/forbidden-path` is better?

joestringer replied:
> I wonder if it would be better to check both allow and reject in a single yaml file since it is the same policy. I.e., do something like: curl ... echo-c/ && ! curl ... echo-c/invalid-path

Single YAML file or single deployment? The idea behind having them report separately in different pods is that we can more easily pinpoint what type of issue it is, rather than "policy doesn't work between these peers" it's "policy *allow* doesn't work between these peers" or "policy *reject* doesn't work between these peers".  That said if we're concerned about deploying too many pods then this could be an easy way to halve the number of these policy-related pods. The trick would be in how we ensure that the `livenessProbe` could easily reflect which of the two errors have occurred.

> Nit: invalid-path was a bit confusing for me because it was not clear that it would be served by the server. Maybe /forbidden-path is better?

Sure, I'll fix this up.

joestringer replied:
> The idea behind having them report separately in different pods is that we can more easily pinpoint what type of issue it is

Another possibility brought up by @tgraf on Slack is to have multiple containers in the same pod.

joestringer replied:
I've reimplemented with 2 containers per pod to test allow/deny L7 paths.

joestringer replied:
Regarding the paths, it's `/public` and `/private` as those are the paths that the underlying server responds to.

kkourt replied:
Thanks! FWIW, the server seems to also respond with empty messages on other paths, but `/public` and `/private` makes it unambiguous.

aanm commented on examples/kubernetes/connectivity-check/connectivity-check-single-node.yaml:
```diff
@@ -268,45 +268,6 @@ spec:
 ---
 apiVersion: apps/v1
 kind: Deployment
-metadata:
```
Is this change intentional in this commit?

joestringer replied:
Kinda, yeah. I mean, I made modifications to the Makefile, then I re-ran `make -C examples/kubernetes/connectivity-check/ clean all` and it removed this section.

Seems to me like from the Makefile, the `single-node` variation in the connectivity check is not supposed to include any hostport checks, and yet the file as checked in has them. I'm guessing we meant to re-generate these YAMLs after an earlier commit but we never did.

I could move this blob out to a separate commit if you think that's meaningful. I didn't want to try to run the above make command before adding the `sort` because I wasn't sure that the file listing order is consistent across different machines.

joestringer replied:
FYI I removed this commit with the new approach, and validated the diff against a deployment of the existing connectivity check. I can provide that diff when I update the PR description.

aanm reviewed (commented):
Minor nit

joestringer committed:
connectivity-check: Add cue CLI tools

Add some basic tooling around connectivity-check YAML generation:

  $ cue cmd help
  List connectivity-check resources specified in this directory

  Usage:
    cue [-t component=<component>] [-t name=<name>] [-t topology=<topology>] <command>

  Available Commands:
    dump   Generate connectivity-check YAMLs from the cuelang scripts
    ls     List connectivity-check resources specified in this directory

List available connectivity-check components:

  $ cue cmd ls
  KIND                  COMPONENT        TOPOLOGY     NAME
  Service               network-check    any          echo-a
  Service               services-check   any          echo-b
  Service               services-check   any          echo-b-headless
  Service               services-check   any          echo-b-host-headless
  Deployment            network-check    any          echo-a
  Deployment            services-check   any          echo-b
  Deployment            services-check   any          echo-b-host
  Deployment            network-check    any          pod-to-a
  Deployment            network-check    any          pod-to-external-1111
  Deployment            policy-check     any          pod-to-a-allowed-cnp
  Deployment            policy-check     any          pod-to-a-denied-cnp
  Deployment            policy-check     any          pod-to-external-fqdn-allow-google-cnp
  Deployment            services-check   multi-node   pod-to-b-multi-node-clusterip
  Deployment            services-check   multi-node   pod-to-b-multi-node-headless
  Deployment            services-check   intra-node   pod-to-b-intra-node-clusterip
  Deployment            services-check   intra-node   pod-to-b-intra-node-headless
  Deployment            services-check   multi-node   host-to-b-multi-node-clusterip
  Deployment            services-check   multi-node   host-to-b-multi-node-headless
  CiliumNetworkPolicy   policy-check     any          pod-to-a-allowed-cnp
  CiliumNetworkPolicy   policy-check     any          pod-to-a-denied-cnp
  CiliumNetworkPolicy   policy-check     any          pod-to-external-fqdn-allow-google-cnp

These can be filtered by component, topology or name. For example:

  $ cue cmd -t component=network ls
  KIND         COMPONENT       TOPOLOGY    NAME
  Service      network-check   any         echo-a
  Deployment   network-check   any         echo-a
  Deployment   network-check   any         pod-to-a
  Deployment   network-check   any         pod-to-external-1111

Finally, to gather the (filtered) YAMLs for the specified resources:

  $ cue cmd dump | head -n 20
  metadata:
    name: echo-a
    labels:
      name: echo-a
      topology: any
      component: network-check
  spec:
    ports:
      - port: 80
    selector:
      name: echo-a
    type: ClusterIP
  apiVersion: v1
  kind: Service
  ---
  ...

Or with an upcoming commit you can just use the Makefile, which now depends
on the cuelang/cue:v0.2.2 Docker image:

  $ make connectivity-check.yaml

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
connectivity-check: Introduce cuelang framework

CUE (https://cuelang.org/) is a data constraint language built defined
as a superset of JSON which aims to "simplify tasks involving defining
and using data".

In context of the connectivity check YAMLs, CUE is useful to allow us to
"evaporate" the boilerplate necessary to define Kubernetes YAMLs for
Deployments, Services and CiliumNetworkPolicies and allow developers to
specify various permutations for connectivity checks concisely.

Why should we use it?

* It's more concise: One template definition, multiple reuse. This is
  useful for introducing new connectivity checks as upcoming commits
  will demonstrate as the developer doesn't need to perform the tedious
  and error-prone process of copying and modifying the YAMLs to
  implement various permutations of a check. Furthermore this helps
  reviewers as they will not have to read through swathes of YAMLs but
  can instead focus on the diffs in the templating that are introduced
  and compare to existing data definitions.

* Consolidate constant declaration. When a core change needs to be made
  to something like the readinessProbe for probes that expect a
  success or failure, we can update one definition in the main CUE file
  and all YAMLs will subsequently be generated with this change in mind.
  During the process of preparing these changes, I noticed
  inconsistencies between various existing YAMLs which appear to just be
  unintentional, where some YAMLs were improved with better timeoute
  behaviour or error rendering, but other YAMLs were left out.

* The data is more structured. Upcoming commits will introduce simple
  CLI tools that allow matching on different classes of connectivity
  checks to generate the corresponding YAMLs. Previously we have
  depended upon file naming schemes and Makefile globbing magic to
  implement this which quickly reaches a limit in which checks should be
  selected for a specific check.

What are the dangers?

* It's relatively immature. At current version v0.2.2 it is subject to
  language changes. Upcoming commits will pin the CLI tool usage to a
  docker container derived from this version to ensure compatibility.
* One more language in the tree to understand, review and interact with.
  Mitigating circumstances: This language comes out of the Golang
  community and as such brings some commonalities; furthermore it is
  beginning to be used in other Kubernetes projects, so there is some
  broader community alignment.
* Its power allows you to hide as much or as little complexity as you
  want. It's tricky to strike a fine balance between explicitly declaring
  (and duplicating) relevant fields in the local file vs. hiding
  convenient templating language in common files. For examples, see
  defaults.cue which automatically derives connectivity check destinations
  based on object name declarations matching regexes of "pod-to-X", and
  applies affinity/anti-affinity via matches on "intra-host" or
  "multi-host".
* All declarations are additive, ie there is no ordering based upon the
  layout in code; instead, data dependencies are determined using the
  declarations, and all data is arranged into a lattice to determine the
  evaluation ordering[0]. This can be counter-intuitive to reason about
  for the uninitiated.

The general approach used in this commit was to `cue import` various
existing YAML files to generate JSON equivalents, then iteratively
combining & consolidating existing definitions using the language
constructs provided by CUE. CUE also provides mechanisms to generate
schemas and autogenerate the structures used here directly from API
definitions (eg from k8s source or Cilium tree), however this area was
not explored in this PR yet. While this doesn't take advantage of one
major aspect of the language, upcoming commits will demonstrate the way
that these changes were validated without the use of standardized
schemas from the underlying Kubernetes resource definitions. (TL;DR:
`kubectl diff ...` with kubectl validation on a live cluster). This was
sufficient to extend the connectivity checks and does not preclude
future explanation of the use of schemas for these definitions.

This commit introduces usage of CUE in a relatively minimal way into the
tree which was useful for my goals of extending the connectivity checks.
If we find that it is useful and powerful, we may consider whether to
extend its usage to other areas of the code (such as for test manifest
generation).

[0] https://cuelang.org/docs/concepts/logic/#the-value-lattice

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
connectivity-check: Support generating YAMLs via cue

Replace the existing YAML generation from individual YAML declarations
for each service with generating YAMLs from the CUE definitions.

Three new targets will assist in validating the migration from the
existing definitions over to CUE:

 * make generate_all
   * For object declared in CUE, generate a file corresponding to that
     definition. For most of the existing YAMLs, this will overwrite the
     copy of the YAML in the tree. This can allow manual inspection of
     individual YAMLs, though the 'inspect' approach is broadly more
     useful for evaluating the overall diff.
 * make deploy
   * Deploy the hostport connectivity checks YAML into an existing cluster.
 * make inspect
   * Generate the YAML file for all connectivity checks, then use
     kubectl to diff these newly generated definitions against the
     running cluster (assuming it was deployed via make deploy).

This commit is purely the makefile changes for easier review &
inspection. Upcoming commits will use these targets to demonstrate that
there is no meaningful change in the generated YAMLs for existing YAMLs
in the tree. In particular, `make inspect` can be used in an iterative
manner by initially deploying the current version of the YAMLs from the
tree, then making changes to the CUE files and inspecting each time a
change is made. When the diff in the cluster represents the changes that
the developer intends to make, the developer can commit the changes to
the CUE files and re-generate the tree versions of the YAMLs.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
connectivity-check: Minor naming fixups

Make some of these resource names a bit more consistent.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
connectivity-check: Replace YAMLs with cue-generated YAMLs

Prior commits introduced CUE definitions that are equivalent to these
YAML files, so we can now:

 * Remove the individual declarations which were previously
   source-of-truth for the connectivity checks
 * Update the overall connectivity-check YAMLs to reflect the minor
   changes that the CUE definitions represent.

To validate this, heavy use of `make inspect` was used. As described in
the prior commit message where this was introduced, this allows diffing
the latest CUE-based YAML definitions against a running copy of the
YAMLs in a cluster.

There are few meaningful changes in this commit which are hard to assess
directly from the git diff, but are easier using `make inspect`:

 * All containers are converted to use readinessProbe and not
   livenessProbe.
 * All readiness probes now specify --connect-timeout of 5s.
 * Readiness probes access `/public` or `/private` per the underlying
   container HTTP server paths rather than just accessing `/`.
 * DNS allow policies are converted to consistently allow both TCP and
   UDP-based DNS.
 * Container names are derived from pod names.
 * The new YAMLs declare additional labels for all resourcess, such as
   'component' and 'topology'.

Signed-off-by: Joe Stringer <joe@cilium.io>

sayboras commented on examples/kubernetes/connectivity-check/echo-c.yaml:
```diff
@@ -0,0 +1,50 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: echo-c
```
nit: conformance test is passed right now, but we might need to add steps to dump details and logs in github action. 
This https://github.com/cilium/cilium/blob/master/.github/workflows/smoke-test.yaml#L153 is my naive attempt, and little verbose.

joestringer replied:
@sayboras is there something we can do from the YAMLs to assist this?

My typical debugging involves basically `kubectl describe [...]` which hopefully tells me which pod + which container is failing, then from there I can try to reason about why that specific check will have failed.

sayboras replied:
Thanks for your offer, but seems unlikely we can do much.

I would love to get to bottom of this https://github.com/cilium/cilium/issues/12279, but it's not failing for my self hosted github action. Current logs we have from github actions didn't help much.

joestringer replied:
I've added a new commit on top that hopes to improve the debugging, and also covers dumping logs for all of the services (by iterating the cue definitions rather than listing them one-by-one).

joestringer commented on examples/kubernetes/connectivity-check/proxy.cue:
```diff
@@ -0,0 +1,80 @@
+deployment: "pod-to-c-multi-node-hostport-proxy-egress": _hostPortProxyResource
+egressCNP: "pod-to-c-multi-node-hostport-proxy-egress": _hostPortProxyPolicy
+deployment: "pod-to-c-intra-node-hostport-proxy-egress": _hostPortProxyResource
+egressCNP: "pod-to-c-intra-node-hostport-proxy-egress": _hostPortProxyPolicy
```
@jrajahalme I've added hostport checks here too.

These lines will generate the YAML to deploy pods with egress L7 policy per the declaration at the top of this file (effectively allowing requests to `/public`), with the destination `"echo-c-host-headless:40001"`, a service that is declared in `echo-servers.cue`. Both intra-node and multi-node checks are added here, with affinity derived from the name of the check here `*-{intra,multi}-node-*` (via `defaults.cue`).

joestringer committed:
connectivity-check: Add hostport + proxy checks

Introduces checks for egress proxy policy when accessing a hostport on a
remote node. These are added as part of the component=hostport-check to
ensure they are not pulled in when running connectivity checks in
environments without hostport support.

Additionally, these new tests are quarantined for now as they are known
to fail in some environments.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
connectivity-check: Add quarantine label to metadata

This new label will be used during YAML generation to ensure that
resources which we are still working on fixes for are kept in a separate
category apart from the regular connectivity checks, to allow us to
check them in & distribute them without causing CI to instantly fail.

Signed-off-by: Joe Stringer <joe@cilium.io>

kkourt commented on examples/kubernetes/connectivity-check/defaults.cue:
```diff
@@ -0,0 +1,40 @@
+		_probeTarget: *"echo-b" | string
+	}
+
+	// Topology
```
AFAIU, the parameters below cannot be overridden, which seems in conflict with the top-level comment.
Maybe a comment here to note the exception makes sense?

joestringer replied:
Ah, true; they need `| string` to be overridden. Could split this section into two: One with fields that may be overwritten and one with final fields.

kkourt commented on examples/kubernetes/connectivity-check/resources.cue:
```diff
@@ -0,0 +1,244 @@
+	}
+
+	if !_allowDNS {
+	spec: egress: _rules
```
nit: missing indentation? (edit: maybe it should be omitted altogether based on following line?)

joestringer replied:
I think I may have only run `cue fmt` on a later commit, can fix this up.

kkourt commented on examples/kubernetes/connectivity-check/services.cue:
```diff
@@ -0,0 +1,38 @@
+package connectivity_check
+
+deployment: [ID=_]: {
```
Any particular reason why this is not placed in `defaults.cue`?

In general, I think that minimizing the places where an entity (in this case `deployment`) is defined might be a good way to strike a good balance between making it easy to write and easy to understand/modify.

joestringer replied:
I tried both approaches, and I didn't conclusively decide one path as superior to the other. Either way, the core definitions are in yet another file (`resources.cue`) and yet in this particular case the definitions are only relevant to this file.

But at the same time, these *will* apply globally so perhaps it's better to shift them into defaults.go. Another option is also to combine defaults.go and resources.go to put all of the code in one place (though I *think* it might be slightly clearer to have defaults.go be smaller and more concise for the type of defaults they provide).

kkourt replied:
I don't feel strongly about either case, so up to you.

How to factor definitions seems like a non-trivial aspect of cue. While the connectivity tests are fairly self-contained and I guess it does not matter much, it's a good playground to experiment and see if and how to extend cue's use.

joestringer replied:
I've added a new commit near the end to document the layout and how to interpret the files.

kkourt reviewed (approved):
Had a look at the CUE-based version, and everything looks great to me!

One suggestion would be to place some parts of the very informative commit messages in a README file in the `connectivity-check` directory.

For example:

- all of the description of how to use the CLI tools from connectivity-check: Add cue CLI tools
- the make targets description from connectivity-check: Support generating YAMLs via cue
- the implementation details from descriptions of  connectivity-check: Introduce proxy checks

joestringer committed:
connectivity-check: Introduce proxy checks

These new checks configure various L7 proxy paths to validate
connectivity via L7 proxies, in the following dimensions:

- Apply policy on egress; ingress; or both (proxy-to-proxy)
- Intra-node / Multi-node
- Allow / Deny

Note that proxy-to-proxy always configures egress allow policy to ensure
that the traffic goes via the proxy and in the drop case the requests
are only rejected at the destination. This is because applying egress
deny at the source would prevent proxy-to-proxy connectivity, meaning
the test would be equivalent to the egress-only reject policy case. This
way, we ensure that the path via the egress proxy to the destination is
tested in the reject case.

These are implemented partially through a new 'echo-c' pod which always
has ingress policy applied to allow GET requests to '/public'. Depending
on whether ingress policy is needed to check the particular permutation
the new checks may connect to 'echo-a' or 'echo-c'.

These are implemented by adding pods for each permutation of policy
apply point and topology; then by adding allow / deny containers within
that pod to test the allow/deny cases.

The 'connectivity-check-proxy.yaml' includes all of the above.

Finally, the omissions: This commit does not attempt to address
variations in datapath configuration. This includes IPv4 vs. IPv6;
tunnel/direct-routing; endpoint config; kube proxy/free; encryption.
These are left up to the cluster operator configuring Cilium in specific
modes and subsequently deploying these YAMLs.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
connectivity-check: Expand readme for latest checks

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
connectivity-check: Re-add liveness probes

It appears that some of these checks require liveness probes rather than
readiness probes to pass on the github actions smoke-test, so ensure all
containers are checked with both.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
smoke-test: Improve state gathering upon failure

Commit bb91571ea497 ("smoke-test: Print pod/deploy state on failure")
attempted to improve the information available during a failure from the
smoke-tests, but only added it to the quick-install test and not the
conformance test. Add the same output also to the conformance test so we
can more easily debug failures there.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
smoke-test: Disable bpf masquerading

In the smoke test, we are relying on kube-proxy for service connectivity
so it doesn't make sense to enable BPF masquerading. In fact, this
causes issues for connectivity from a node to a pod on a remote node via
ClusterIP (see related issue).

For the moment, disable BPF masquerading while we figure out the
longer-term solution to that issue.

Related: #12699

Signed-off-by: Joe Stringer <joe@cilium.io>

errordeveloper reviewed (commented):
Didn't get a chance to have an in-depth look, but generally I think that using CUE for this is a very good idea, we should consider using it more in various testing use-cases. I also found that driving CUE from Go is fairly easy, we are doing that in the GKE test cluster operator, it's also more easily unit-testable that way.

sayboras commented on .github/workflows/smoke-test.yaml:
```diff
@@ -139,6 +133,7 @@ jobs:
              --set global.externalIPs.enabled=true \
              --set global.nodePort.enabled=true \
              --set global.hostPort.enabled=true \
+             --set config.bpfMasquerade=false \
```
:+1:

sayboras replied:
nit: this one should be added in https://docs.cilium.io/en/latest/gettingstarted/kind as well.

joestringer replied:
I've noted that down on issue #12699 to follow up.

sayboras commented on examples/kubernetes/connectivity-check/Makefile:
```diff
@@ -1,35 +1,90 @@
+CUE := $(DOCKER_RUN) $(CUE_IMAGE)
 
-$(DEFAULT_OUT): $(DEFAULT_SRC)
+all: $(ALL_TARGETS)
```
nit: might need to add step to run this target and check `git diff` before running smoke test. Can be done as follow up.

joestringer replied:
Filed https://github.com/cilium/cilium/issues/12714.

sayboras commented on examples/kubernetes/connectivity-check/connectivity-check.yaml:
```diff
@@ -1,112 +1,147 @@
-kind: Deployment
 metadata:
   name: echo-a
+  labels:
```
nit: we might need to update docs related to connectivity check as there are some name changes, and scope changes.

Can be done later.

joestringer replied:
New commit incoming for this.

sayboras commented on examples/kubernetes/connectivity-check/Makefile:
```diff
@@ -1,35 +1,90 @@
 
-all: $(DEFAULT_OUT) $(HOSTPORT_OUT) $(SINGLE_OUT)
+CUE_IMAGE := "docker.io/cuelang/cue:v0.2.2@sha256:2ea932c771212db140c9996c6fff1d236f3f84ae82add914374cee553b6fc60c"
+CUE := $(DOCKER_RUN) $(CUE_IMAGE)
```
nit: $(QUIET) can be added to reduce verbosity, e.g. make help

joestringer replied:
FWIW I'd need to add $(QUIET) to the individual make lines rather than in the definition here, just so that it won't get messed up if we call cue within a loop.

sayboras commented on examples/kubernetes/connectivity-check/Makefile:
```diff
@@ -1,35 +1,90 @@
+	@rm -f *.yaml
+
+deploy:
+	kubectl apply -f connectivity-check-hostport.yaml
```
should it be common `connectivity-check.yaml` ?

joestringer replied:
Good question. Here I opted for the larger YAML to try to ensure that any developers who use this target will get more information (even checks that may potentially fail depending on cluster configuration), just to reduce the likelihood that we overlook the extra checks in the hostport YAML.

sayboras reviewed (approved):
I don't know cuelang, so review generated files only.

LGTM :100:

nebril commented on examples/kubernetes/connectivity-check/defaults.cue:
```diff
@@ -0,0 +1,63 @@
+// Default parameters for echo clients (final).
+deployment: [ID=_]: {
+	// Topology
+	if ID =~ "^[-_a-zA-Z0-9]*intra-node[-_a-zA-Z0-9]*$" {
```
question - any particular reason why you are limiting the character set? If it's for render-time validation I think that's fine, but maybe we should validate these yamls via some kind of local validator as a build step?

joestringer replied:
Maybe it can just be `.*` instead, I don't have a specific motivation to restrict the characterset.

Is it possible to validate locally? I have been using the standard kubectl validation (which requires the resources such as CNP to already be configured in the connected cluster with schema validation included).

nebril reviewed (approved):
Looks great! 

Linter failed for some reason in `pkg/k8s/apis/cilium.io/v2/types.go:41`, not sure if it's connected to this PR.

joestringer committed:
docs: Update connectivity-check examples

Signed-off-by: Joe Stringer <joe@cilium.io>

nathanjsweet commented on examples/kubernetes/connectivity-check/README.md:
```diff
@@ -2,3 +2,35 @@
+multiple files per the following logic:
+
+* `resources.cue`: The main definitions for templating all Kubernetes resources,
+  including Deployment, Service and CiliumNetworkPolicy.
```
Oxford (or serial) comma is usually deemed standard. change to: `Deployment, Service and, CiliumNetworkPolicy`

joestringer replied:
Is the comma after `resources` supposed to be there?

nathanjsweet commented on examples/kubernetes/connectivity-check/README.md:
```diff
@@ -2,3 +2,35 @@
+* `echo-servers.cue`: Data definitions for all `echo-*` servers used for other
+  connectivity checks.
+* `defaults.cue`: Default parameters used to define how specific checks connect
+  to particular echo servers, including selecting the probe destination;
```
I'm finding this semi-colon confusing. Could it be replace with an "and"?

joestringer replied:
it's poorly placed :-) "x and y and z" is an unusual way to list things though. I think this is another job for the Oxford comma.

nathanjsweet reviewed (changes requested):
Two nits. Otherwise, it looks good.
